ext/drx/drx_buf.c

```diff
@@ -648,6 +649,96 @@ drx_buf_insert_buf_store(void *drcontext, drx_buf_t *buf, instrlist_t *ilist,
     }
 }
 
+static void
+insert_load(void *drcontext, instrlist_t *ilist, instr_t *where,
+                    reg_id_t dst, reg_id_t src, opnd_size_t opsz)
+{
+    switch (opsz) {
+    case OPSZ_1:
+        MINSERT(ilist, where, XINST_CREATE_load_1byte
+                (drcontext,
+                 opnd_create_reg(reg_resize_to_opsz(dst, opsz)),
+                 opnd_create_base_disp(src, DR_REG_NULL, 0, 0, opsz)));
+        break;
+    case OPSZ_2:
+        MINSERT(ilist, where, XINST_CREATE_load_2bytes
+                (drcontext,
+                 opnd_create_reg(reg_resize_to_opsz(dst, opsz)),
+                 opnd_create_base_disp(src, DR_REG_NULL, 0, 0, opsz)));
+        break;
+    case OPSZ_4:
+#if defined(X86_64) || defined(AARCH64)
+    case OPSZ_8:
+#endif
+        MINSERT(ilist, where, XINST_CREATE_load
+                (drcontext,
+                 opnd_create_reg(reg_resize_to_opsz(dst, opsz)),
+                 opnd_create_base_disp(src, DR_REG_NULL, 0, 0, opsz)));
+        break;
+    default:
+        DR_ASSERT(false);
+        break;
+    }
+}
+
+DR_EXPORT
+void
+drx_buf_insert_buf_memcpy(void *drcontext, drx_buf_t *buf, instrlist_t *ilist,
+                          instr_t *where, reg_id_t dst, reg_id_t src, ushort len)
+{
+    opnd_size_t opsz = opnd_size_from_bytes(len);
+
+    if (len > sizeof(app_pc)) {
+        /* slow path, we call into our personal memcpy() implementation */
+        /* XXX: We should elect to write our own memcpy() built in DR IR. This
+         * would potentially be slower than the glibc memcpy, but then we can
+         * incrementally update the buffer pointer so the user can recover
+         * from partial writes.
+         * XXX: It's also unlikely that if glibc's memcpy faults, we'd be able
+         * to recover from it -- an in-app fault will not be handled by us, and
+         * also the buffer pointer is not updated incrementally.
```

Comment:
> Another thought occurred to me: instead of going through all this, would it maybe be possible to call dr_insert_clean_call() on a routine which performs a DR_TRY_EXCEPT to memcpy into the user's buffer? If we fault we can invoke the user's callback, reset the user's buffer pointer and then retry the write.

DR_TRY_EXCEPT has higher overhead as it saves the context ahead of time, every time, even when there's no fault.